feat: add GET `/api/v1/me/workspaces` & update `Workspace` class

alvarobartt opened the pull request:
# Description

This PR adds the `/api/v1/me/workspaces` endpoint in the API V1, and also updates the `Workspace` accordingly so that both users with role annotator and/or admin, can list the workspaces they are part of.

So on, the bug related to the access to the `User.workspaces` property is now solved, as before just the `owners` were able to access it and print the model (as `workspace` was part of the repr).

Besides that, some minor things have also been tackled as part of this PR:
* Remove unused imports
* Restrict `Workspace.delete_user` to users with role `owner` only

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [X] Add unit tests for GET `/api/v1/me/workspaces`
- [X] Add unit tests for `list_workspaces_me` in Python SDK
- [X] Add unit tests for `User.workspaces` in Python client

All the above tested for all the roles: owner, admin, and annotator.

**Checklist**

- [X] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

alvarobartt committed:
fix: remove `allowed_for_roles` for `User.workspaces`

alvarobartt committed:
chore: remove unused imports

alvarobartt committed:
feat: add `list_workspaces_me` for every user role

alvarobartt committed:
feat: add `list_workspaces_me` in SDK

alvarobartt committed:
fix: `User.workspaces` to work for `admin` & `annotator`

alvarobartt committed:
fix: restrict `admin` from `Workspace.delete_user`

alvarobartt committed:
test: `test_user_repr` to check every role

alvarobartt committed:
test: add `list_workspaces_me` unit tests

alvarobartt committed:
chore: remove unused imports

alvarobartt committed:
chore: remove unused import

alvarobartt committed:
test: add `/api/v1` SDK unit tests for workspaces

alvarobartt committed:
docs: update `CHANGELOG.md`

frascuchon commented on src/argilla/client/users.py:
```diff
@@ -108,14 +109,16 @@ def __init__(
+        if self.role == UserRole.owner:
+            return users_api_v1.list_user_workspaces(self.__client, self.id).parsed
+        else:
+            return workspaces_api_v1.list_workspaces_me(self.__client).parsed
```
this is working because the admin and annotator role cannot create `User` without using `User.me`, right?

frascuchon replied:
I mean, ideally, to launch the `list_workspaces_me` the condition should verify that `self` is the connected user. Anyway, this should work

alvarobartt replied:
No, you're right, indeed if we list users from the `owner` we won't get the real results, I'll add the corrected statement now! Nice catch 😄

frascuchon replied:
This change will have a problem if the client is reinitialized after creating the user instance. We should find another approach

alvarobartt committed:
fix: `User.workspaces` to check active role

When checking `self.role` we were checking the user's instance role, not the active user one

alvarobartt committed:
test: add `test_user_workspaces_from_owner_to_any`

frascuchon commented on src/argilla/client/users.py:
```diff
@@ -108,14 +109,16 @@ def __init__(
+        if active_client().user.role == UserRole.owner:
+            return users_api_v1.list_user_workspaces(self.__client, self.id).parsed
+        else:
+            return workspaces_api_v1.list_workspaces_me(self.__client).parsed
```
```suggestion
        # TODO: improve this
        connected_user = users_api.whoami_httpx(self.__client).parsed

        if connected_user.role == UserRole.owner:
            return users_api_v1.list_user_workspaces(self.__client, self.id).parsed

        elif connected_user.id == self.id:
            return workspaces_api_v1.list_workspaces_me(self.__client).parsed
        else:
            raise PermissionError(
                f"User with role={connected_user.role} is not allowed to workspaces for user `{self.id}`."
            )
```

frascuchon replied:
Or just skip the id check:
```suggestion
        # TODO: improve this
        connected_user = users_api.whoami_httpx(self.__client).parsed

        if connected_user.role == UserRole.owner:
            return users_api_v1.list_user_workspaces(self.__client, self.id).parsed
        return workspaces_api_v1.list_workspaces_me(self.__client).parsed
```

alvarobartt replied:
We can skip indeed the `user_id` check because any other user than `owner` is not able to use any other method besides `User.me()` initially

alvarobartt replied:
And regarding the proposal, shouldn't we use the current active user instead of an HTTP request? This way we'll avoid sending 2 requests to Argilla on every call to `User.workspaces`

alvarobartt replied:
I assume the use-case is changes in the role of the current user, which would be problematic if we're using an `admin` or `annotator`, as we can just use the `User.me` classmethod, and changing the role of the active user will still make the class invalid, as the `role` attribute won't be updated, so IMO there's no need to send that extra HTTP request

frascuchon replied:
Yes, it's why I put the TODO to improve this. My concern about your proposal is regarding the use of `active_client` inside this method. This client can be misaligned with the `__client` when the user instance was created, so we can use the wrong info to check the workspaces access

frascuchon commented on src/argilla/server/apis/v1/handlers/workspaces.py:
```diff
@@ -44,3 +44,15 @@ async def get_workspace(
+    await authorize(current_user, WorkspacePolicyV1.list_workspaces_me)
+
+    workspaces = await accounts.list_workspaces_by_user_id(db, current_user.id)
+    return Workspaces(items=workspaces)
```
```suggestion
    await authorize(current_user, WorkspacePolicyV1.list_workspaces_me)

    if current_user.is_owner:
        workspaces = await accounts.list_workspaces(db)
    else:
        workspaces = await accounts.list_workspaces_by_user_id(db, current_user.id)
    return Workspaces(items=workspaces)
```

alvarobartt replied:
Thanks for the suggestion (sorry I forgot to include you as co-author 😞)

frascuchon commented on tests/server/api/v1/test_workspaces.py:
```diff
@@ -75,3 +75,40 @@ async def test_get_workspace_with_nonexistent_workspace_id(client: TestClient, o
+            "name": workspace.name,
+            "inserted_at": workspace.inserted_at.isoformat(),
+            "updated_at": workspace.updated_at.isoformat(),
+        } in response.json()["items"]
```
```suggestion
@pytest.mark.asyncio
async def test_list_workspaces_me_as_owner(client: TestClient, owner_auth_header: dict) -> None:
    workspaces = await WorkspaceFactory.create_batch(size=5)

    response = client.get("/api/v1/me/workspaces", headers=owner_auth_header)

    assert response.status_code == 200
    assert len(response.json()["items"]) == len(workspaces)
    for workspace in workspaces:
        assert {
            "id": str(workspace.id),
            "name": workspace.name,
            "inserted_at": workspace.inserted_at.isoformat(),
            "updated_at": workspace.updated_at.isoformat(),
        } in response.json()["items"]
 
               
@pytest.mark.asyncio
@pytest.mark.parametrize("role", [UserRole.annotator, UserRole.admin])
async def test_list_workspaces_me(client: TestClient, role: UserRole) -> None:
    workspaces = await WorkspaceFactory.create_batch(size=5)
    user = await UserFactory.create(role=role, workspaces=workspaces)

    response = client.get("/api/v1/me/workspaces", headers={API_KEY_HEADER_NAME: user.api_key})

    assert response.status_code == 200
    assert len(response.json()["items"]) == len(workspaces)
    for workspace in workspaces:
        assert {
            "id": str(workspace.id),
            "name": workspace.name,
            "inserted_at": workspace.inserted_at.isoformat(),
            "updated_at": workspace.updated_at.isoformat(),
        } in response.json()["items"]
```

frascuchon commented on tests/client/sdk/v1/workspaces/test_api.py:
```diff
@@ -0,0 +1,51 @@
+    assert isinstance(response.parsed, list)
+    assert len(response.parsed) > 0
+    assert isinstance(response.parsed[0], WorkspaceModel)
+    assert len(response.parsed) == len(workspaces)
```
```suggestion
@pytest.mark.asyncio
async def test_list_workspaces_me_as_owner(owner: User) -> None:
    workspaces = await WorkspaceFactory.create_batch(size=5)

    httpx_client = ArgillaSingleton.init(api_key=user.api_key).http_client.httpx

    response = list_workspaces_me(client=httpx_client)
    assert response.status_code == 200
    assert isinstance(response.parsed, list)
    assert len(response.parsed) > 0
    assert isinstance(response.parsed[0], WorkspaceModel)
    assert len(response.parsed) == len(workspaces)

@pytest.mark.asyncio
@pytest.mark.parametrize("role", [UserRole.admin, UserRole.annotator])
async def test_list_workspaces_me(role: UserRole) -> None:
    workspaces = await WorkspaceFactory.create_batch(size=5)
    user = await UserFactory.create(role=role, workspaces=workspaces)

    httpx_client = ArgillaSingleton.init(api_key=user.api_key).http_client.httpx

    response = list_workspaces_me(client=httpx_client)
    assert response.status_code == 200
    assert isinstance(response.parsed, list)
    assert len(response.parsed) > 0
    assert isinstance(response.parsed[0], WorkspaceModel)
    assert len(response.parsed) == len(workspaces)
```

alvarobartt committed:
fix: call `list_workspaces` if `is_owner=True`

alvarobartt committed:
fix: `list` and `from_name` to use `list_workspaces_me`

alvarobartt committed:
feat: temporarily add `whoami_httpx`

alvarobartt committed:
fix: `User.workspaces` relative to `__client`

alvarobartt committed:
docs: add `TODO` in `whoami_httpx`

alvarobartt committed:
fix: `allowed_for_roles` to use `whoami_httpx`

alvarobartt committed:
fix(test): assign workspaces before listing

alvarobartt committed:
test: `list_workspaces_me` with all roles

alvarobartt committed:
fix(test): assign `workspaces` to `admin` & `annotator` only

alvarobartt committed:
fix(docs): add `role` in `User` docstring

alvarobartt committed:
revert: remove `workspaces` from `__repr__`
